gk/fib.c

```diff
@@ -0,0 +1,1640 @@
+	 */
+	if (ip_prefix_fib->u.grantor.next_fib->action ==
+			GK_FWD_NEIGHBOR_BACK_NET) {
+		/* Find the FIB entry for the Grantor. */
+		struct neighbor_hash_table *neigh_ht;
+
+		if (flow->proto == ETHER_TYPE_IPv4 && flow->f.v4.dst ==
+				ip_prefix_addr->ip.v4.s_addr) {
+			/*
+			 * The Ethernet header cache entry is still
+			 * referenced by another prefix, so we cannot
+			 * release this FIB entry.
+			 */
+			if (eth_cache->ref_cnt != 1) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: delete a grantor neighbor fib entry [ip prefix = %s], however, its cached Ethernet header is referenced by other entries\n",
+					ip_prefix);
+				return -1;
+			}
+
+			/*
+			 * Find the neighbor hash table that
+			 * that contains its @eth_cache.
+			 */
+			neigh_ht = &ip_prefix_fib->u.
+				grantor.next_fib->u.neigh;
+			ret = neigh_del_ether_cache(neigh_ht,
+				&flow->f.v4.dst);
+			if (ret < 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: failed to delete the Ethernet cached header of the Grantor FIB entry for the IP prefix %s at %s\n",
+					ip_prefix, __func__);
+				return -1;
+			}
+		} else if (flow->proto == ETHER_TYPE_IPv6 &&
+				memcmp(flow->f.v6.dst,
+				ip_prefix_addr->ip.v6.s6_addr,
+				sizeof(ip_prefix_addr->ip.v6) == 0)) {
+			/*
+			 * The Ethernet header cache entry is still
+			 * referenced by another prefix, so we cannot
+			 * release this FIB entry.
+			 */
+			if (eth_cache->ref_cnt != 1) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: delete a grantor neighbor fib entry [ip prefix = %s], however, its cached Ethernet header is referenced by other entries\n",
+					ip_prefix);
+				return -1;
+			}
+
+			/*
+			 * Find the neighbor hash table that
+			 * that contains its @eth_cache.
+			 */
+			neigh_ht = &ip_prefix_fib->u.
+				grantor.next_fib->u.neigh6;
+			ret = neigh_del_ether_cache(neigh_ht,
+				flow->f.v6.dst);
+			if (ret < 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: failed to delete the Ethernet cached header of the Grantor FIB entry for the IP prefix %s\n",
+					ip_prefix);
+				return -1;
+			}
+		} else {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: remove an invalid FIB entry for a Grantor with IP type %hu at %s",
+				flow->proto, __func__);
+			return -1;
+		}
+	} else {
+		/*
+		 * The Grantor is not a neighbor, we need to decrement
+		 * the @eth_cache->ref_cnt.
+		 */
+		eth_cache->ref_cnt--;
+	}
+
+	return 0;
+}
+
+/*
+ * This function is called by del_fib_entry_locked().
+ * Notice that, it doesn't stand on its own, and it's only
+ * a construct to make del_fib_entry_locked() readable.
+ */
+static int
+del_gateway_fib_entry_locked(
+	const char *ip_prefix, struct ipaddr *ip_prefix_addr,
+	struct gk_fib *ip_prefix_fib, struct gk_config *gk_conf)
+{
+	int ret;
+	struct ipaddr *gw_addr =
+		&ip_prefix_fib->u.gateway.ip_addr;
+	struct gk_fib *neigh_fib;
+	struct neighbor_hash_table *neigh_ht;
+	struct ether_cache *eth_cache =
+		ip_prefix_fib->u.gateway.eth_cache;
+
+	/*
+	 * The @ip_prefix is a gateway.
```

Comment:
It is more precise to change `@ip_prefix` to `@ip_prefix_fib`. Please search for other locations where similar comments are being used.